conans/client/cmd/uploader.py

```diff
@@ -253,22 +272,50 @@ def upload_package(self, pref, prev_bundle, remote):
         output.debug(f"Upload {pref} in {duration} time")
 
 
-def compress_files(files, name, dest_dir, compresslevel=None, ref=None):
+def compress_files(files, name, dest_dir, compressformat=None, compresslevel=None, ref=None):
     t1 = time.time()
-    # FIXME, better write to disk sequentially and not keep tgz contents in memory
-    tgz_path = os.path.join(dest_dir, name)
+    tar_path = os.path.join(dest_dir, name)
     ConanOutput(scope=str(ref)).info(f"Compressing {name}")
-    with set_dirty_context_manager(tgz_path), open(tgz_path, "wb") as tgz_handle:
-        tgz = gzopen_without_timestamps(name, mode="w", fileobj=tgz_handle,
-                                        compresslevel=compresslevel)
-        for filename, abs_path in sorted(files.items()):
-            # recursive is False in case it is a symlink to a folder
-            tgz.add(abs_path, filename, recursive=False)
-        tgz.close()
+
+    if compressformat == "zstd":
+        with open(tar_path, "wb") as tarfile_obj:
+            # Only provide level if it was overridden by config.
+            zstd_kwargs = {}
+            if compresslevel is not None:
+                zstd_kwargs["level"] = compresslevel
+
+            dctx = zstandard.ZstdCompressor(write_checksum=True, threads=-1, **zstd_kwargs)
+
+            # Create a zstd stream writer so tarfile writes uncompressed data to
+            # the zstd stream writer, which in turn writes compressed data to the
+            # output tar.zst file.
+            with dctx.stream_writer(tarfile_obj) as stream_writer:
+                # The choice of bufsize=32768 comes from profiling compression at various
+                # values and finding that bufsize value consistently performs well.
+                # The variance in compression times at bufsize<=64KB is small. It is only
+                # when bufsize>=128KB that compression times start increasing.
+                with tarfile.open(mode="w|", fileobj=stream_writer, bufsize=32768,
+                                  format=tarfile.PAX_FORMAT) as tar:
+                    unflushed_bytes = 0
+                    for filename, abs_path in sorted(files.items()):
+                        tar.add(abs_path, filename, recursive=False)
+
+                        unflushed_bytes += os.path.getsize(abs_path)
+                        if unflushed_bytes >= 2097152:
```

Comment:
You are correct, it should have been frames, not blocks. Frames are where the decoder can arbitrarily restart, while blocks within a frame have a data dependency in the uncompressed stream. (Therefor compression can be trivially threaded, but decompression can't.)

Just enforcing frames does not provide any parallel decompression capabilities with the pure zstd library out of the box either, unfortunately, so still only compression can be threaded. (I would have referenced `pzstd` as an example of how to multithread decompression as well - but it's on a deprecation path and did actually require skippable non-standard frames in between the compressed frames in order to actually achieve multi-threaded decompression from multiple offsets of the input file. Without those extra index frames, you still can't find the end of a frame without having to scan the file first.)

Where enforced framing is useful though, is in 3rd party tools (e.g. Artifactory!) that need to provide pseudo random access which doesn't work at all with the fully streamed default format. I know conan center doesn't care, but packages with debug symbols for some of the larger recipes tend to go into the multi GB range, and having to decode the full file to serve something from in the middle (e.g. because you have a symbol server that's pointing straight into the (indexed) archive contents in an Artifactory!) is simply madness.

2MB is way too little though. ZSTD is realistically designed to process 50MB+ in a single streaming pass, with the ***highest*** compression levels reaching up to an 128MB window size. Realistically, you would only want to enforce a frame break at around 64-128MB, as this ensures that the loss in compression ratio is still minimal, but "random access" is still reasonably (as in: up to a 50-100ms of additional CPU time, rather than 10s or more) costly.